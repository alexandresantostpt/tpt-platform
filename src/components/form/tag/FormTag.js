import { events } from '@constants/events'

import PropTypes from 'prop-types'
import PubSub from 'pubsub-js'
import React, { useEffect, useState } from 'react'

import TagInputContainer from './container/TagInputContainer'
import TagInputItem from './item/TagInputItem'
import TagInputList from './list/TagInputList'

import CitiesSuggest from '@components/form/suggest/CitiesSuggest'

const FormTag = props => {
    const { disabled, isInvalid, resetSuggestions } = props
    const [tags, updateTags] = useState([])

    useEffect(() => {
        PubSub.subscribe(events.FORM_RESET, reset)
        return () => PubSub.unsubscribe(events.FORM_RESET, reset)
    }, [])

    useEffect(() => {
        const { initialValue } = props
        if (initialValue.length) {
            updateTags([...initialValue])
        }
        return reset
    }, [props.initialValue])

    const handleAddTag = citySelected => {
        const { form, name } = props

        if (citySelected) {
            form.current.form.change(name, [...tags, citySelected])
            updateTags([...tags, citySelected])
        }
    }

    const handleRemoveTag = index => {
        const { form, name } = props
        const newTags = tags.filter((_, i) => i !== index)
        form.current.form.change(name, newTags)
        updateTags(newTags)
    }

    const reset = () => {
        updateTags([])
    }

    return (
        <>
            <TagInputList>
                {tags.map((tag, i) => (
                    <TagInputItem disabled={disabled} key={tag.name} onClick={() => handleRemoveTag(i)}>
                        {tag.name}
                    </TagInputItem>
                ))}
            </TagInputList>
            <TagInputContainer>
                <CitiesSuggest
                    {...props}
                    addTag={handleAddTag}
                    disabled={disabled}
                    isInvalid={isInvalid}
                    resetSuggestions={resetSuggestions}
                />
            </TagInputContainer>
        </>
    )
}

FormTag.propTypes = {
    disabled: PropTypes.bool.isRequired,
    form: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    initialValue: PropTypes.oneOfType([PropTypes.array, PropTypes.string]).isRequired,
    isInvalid: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    onBlur: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    required: PropTypes.bool.isRequired,
    type: PropTypes.string.isRequired
}

export default FormTag
