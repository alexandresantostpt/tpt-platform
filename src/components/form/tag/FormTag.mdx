---
menu: FormTag
name: FormTag
route: /components/form/FormTag
---

import { Playground, PropsTable } from 'docz'

import FormTag from './FormTag'

import '@cfg/scripts'
import '@cfg/style'

# FormTag

<PropsTable of={FormTag} />

## Basic usage

```
<FormTag
    form={ FORM_REF }
    id="MY_INPUT"
    name="MY_INPUT"
    onBlur={ () => console.log('Blur...') }
    onChange={ () => console.log('Change...') }
    onFocus={ () => console.log('Focus...') }
    required={ true }
    type="text"
    value="MY_VALUE"
/>
```

<Playground>
    <FormTag
        form={{}}
        id="MY_INPUT"
        name="MY_INPUT"
        onBlur={() => console.log('Blur...')}
        onChange={() => console.log('Change...')}
        onFocus={() => console.log('Focus...')}
        required={true}
        type="text"
        value="MY_VALUE"
    />
</Playground>
