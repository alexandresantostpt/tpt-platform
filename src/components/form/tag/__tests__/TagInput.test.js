import React from 'react'
import renderer from 'react-test-renderer'

import FormTag from '../FormTag'

describe('Tests for TagInput component', () => {
    it('Should render without crash', () => {
        const tree = renderer
            .create(
                <FormTag
                    form={{}}
                    id="MY_INPUT"
                    name="MY_INPUT"
                    onBlur={() => console.log('Blur...')}
                    onChange={() => console.log('Change...')}
                    onFocus={() => console.log('Focus...')}
                    required={true}
                    type="text"
                    value="MY_VALUE"
                />
            )
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
