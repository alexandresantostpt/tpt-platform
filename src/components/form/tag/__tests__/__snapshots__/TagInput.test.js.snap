// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for TagInput component Should render without crash 1`] = `
Array [
  <ul
    className="TagInputList__StyledList-sc-10dh3l6-0 fIqMKr"
  />,
  <div
    className="TagInputContainer__StyledInputContainer-qji6kn-0 iBoEuw"
  >
    <div
      className="CitiesSuggest__StyledAutosuggest-sc-15ojzwf-0 jIAIbN"
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-owns="react-autowhatever-MY_INPUT"
        className="react-autosuggest__container"
        role="combobox"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls="react-autowhatever-MY_INPUT"
          autoComplete="off"
          className="react-autosuggest__input"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="text"
          value=""
        />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-MY_INPUT"
          role="listbox"
        />
      </div>
    </div>
  </div>,
]
`;
